Refactor FinancialDatasetAPI integration and the memory cache

anhquan075 opened the pull request:
This pull request focuses on refactoring the codebase to use a new `FinancialDatasetAPI` class for accessing financial data, replacing individual API functions. The changes involve using this new class, removing the old API functions, and implementing a base API client class for better structure and error handling.

### Refactoring to use `FinancialDatasetAPI`:

* [`src/agents/market_data.py`](diffhunk://#diff-c9a7f096719250382ab7f769dcd5d574c0ed092708507d17da1c82f5223dcd9aR1-R73): Replaced individual API function calls with methods from the new `FinancialDatasetAPI` class. Adjusted the import statements accordingly.
* `src/agents/risk_manager.py`, `src/agents/technicals.py`, `src/backtester.py`: Updated to use `FinancialDatasetAPI.prices_to_df` for converting price data to a DataFrame. 

### Removal of old API functions:

* [`src/tools/api.py`](diffhunk://#diff-e20f3acd8c0a3a5f227c9942aa5333700a227afd267b10fe5a02495da8547856L1-L152): Removed all old API functions such as `get_financial_metrics`, `search_line_items`, `get_insider_trades`, `get_market_cap`, `get_prices`, and `prices_to_df`.

### Implementation of base API client:

* [`src/tools/api/base.py`](diffhunk://#diff-25c23803948fed10d2b0f82895ce40a17e965640cd7f08baca542c71ffa0e23aR1-R60): Made an abstract base class `BaseAPIClient` to handle common API client functionality such as making requests and handling responses.
* [`src/tools/api/config.py`](diffhunk://#diff-3afe8d3f042beccc62c28f80fc047ef2231a8c16b3f5b9487d1d6c1f40cdd42dR1-R31): Added configuration classes `BaseAPIConfig` and `FinancialDatasetAPIConfig` for managing API configuration.
* [`src/tools/api/financial_dataset.py`](diffhunk://#diff-868e2f453bb81dbc7ed3e4ab5a60f885627c01ab63c5d2cc13754f1ffbd9c88eR1-R185): Created the `FinancialDatasetAPI` class extending `BaseAPIClient`, implementing methods for fetching financial metrics, insider trades, market cap, and price data.

hoainam-nguyen commented:
Awesome

marc-at-brightnight commented on src/agents/market_data.py:
```diff
@@ -1,74 +1,89 @@
-            **data, 
-            "prices": prices, 
-            "start_date": start_date, 
+            **data,
```
3.9+ syntax is preferred, i.e. `data | { ... }`, rather than `{ **data, ... }`

marc-at-brightnight commented on src/tools/api/financial_dataset.py:
```diff
@@ -0,0 +1,185 @@
+from .config import FinancialDatasetAPIConfig
+
+
+class Period(str, Enum):
```
don't think you need `str`

marc-at-brightnight commented on src/tools/api/financial_dataset.py:
```diff
@@ -0,0 +1,185 @@
+    def _get_headers(self) -> Dict[str, str]:
+        return {"X-API-KEY": self.config.api_key}
+
+    def _handle_response(self, response: requests.Response) -> Dict[str, Any]:
```
use `dict` instead of `Dict`. I would add a type alias for `dict[str, Any]` in an appropriate file, since it's used so often (arguably too often)

marc-at-brightnight commented on src/tools/api/financial_dataset.py:
```diff
@@ -0,0 +1,185 @@
+
+        return df.sort_index()
+
+    def get_financial_metrics(
```
public methods usually come before private methods

anhquan075 replied:
Because the `get_financial_metrics` method is already used as a public function in some files, I will maintain it for consistency.

marc-at-brightnight commented on src/tools/api/financial_dataset.py:
```diff
@@ -0,0 +1,185 @@
+        return response.json()
+
+    @staticmethod
+    def prices_to_df(prices: List[Dict[str, Any]]) -> pd.DataFrame:
```
use `list` instead of `List`

marc-at-brightnight commented on src/tools/api/financial_dataset.py:
```diff
@@ -0,0 +1,185 @@
+        )
+
+    def get_insider_trades(
+        self, ticker: str, end_date: Union[str, datetime], limit: int = 5
```
instead of `Union`, use `str | datetime`

marc-at-brightnight reviewed (commented):
Nice work 👏🏼  Would probably just recommend some typing improvements

mripani commented:
Heads up for this PR! 

This will allow to easily integrate with other API data providers which is needed for the project.

anhquan075 commented:
Hi team, thanks for your patience. I'm just back from the public holiday and have now reviewed and addressed all comments, as well as merged with the latest branch. I appreciate your attention to this PR and hope you can take some time to help me review it.

anhquan075 commented:
Hi @virattt , any chance for me to test the API behavior without using the API key? 🫠 I’m not sure if the minimum top-up credit will be enough for my needs.

dtiukalov commented:
since you introduced `_process_price_data()`, calling `prices_to_df()` is redundant (in `technicals.py` and `risk_manager.py`)

dtiukalov commented on src/agents/fundamentals.py:
```diff
@@ -17,18 +18,22 @@ def fundamentals_agent(state: AgentState):
-        financial_metrics = get_financial_metrics(
+        financial_metrics = FinancialDatasetAPI().get_financial_metrics(
             ticker=ticker,
             end_date=end_date,
```
`report_period=end_date,`

dtiukalov commented on src/agents/valuation.py:
```diff
@@ -13,29 +14,33 @@ def valuation_agent(state: AgentState):
-        financial_metrics = get_financial_metrics(
+        financial_metrics = financial_api.get_financial_metrics(
             ticker=ticker,
             end_date=end_date,
```
report_period=end_date,

dtiukalov commented:
in `risk_manager.py` shall be `from tools.api import FinancialDatasetAPI`

dtiukalov commented:
check everywhare: instead of `if not prices:` -> `if prices.empty:`

dtiukalov commented:
using `import fireducks.pandas as pd` instead of `import pandas as pd` may improve performance

anhquan075 commented:
> using `import fireducks.pandas as pd` instead of `import pandas as pd` may improve performance

It seems the `fireducks` isn't supported for Mac ARM rn @dtiukalov, so I can't test with this framework.

duke-cliff commented on src/tools/api/financial_dataset.py:
```diff
@@ -0,0 +1,316 @@
+        """
+        Search for specific line items in financial statements.
+        """
+        cache_key = f"line_items:{ticker}_{period}_{'-'.join(sorted(line_items))}"
```
should store in multiple cache_keys, one per item
